modules/c++/xml.lite/source/Speed.cpp

```diff
@@ -0,0 +1,281 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this program; If not,
+ * see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+#include <string>
+#include <vector>
+#include <map>
+#include <stdio.h>
+#include <stdlib.h>
+#include <import/str.h>
+#include <import/except.h>
+#include "xml/lite/Element.h"
+#include "xml/lite/Speed.h"
+
+//=================================================================
+void xml::lite::Speed::fromDocument(xml::lite::Document *doc, bool steal)
+{
+  setRootElement(doc->getRootElement(steal), steal);
+  buildTOC();
+}
+//=================================================================
+void xml::lite::Speed::buildTOC(const std::string &parent,
+                         xml::lite::Element const * element)
+{
+  // make a smaller, local map for bookkeeping.
+  std::map<std::string, size_t> local_registry;
+  // either both or niether parent and element must be specified
+  if (parent.empty() ^ (element == NULL))
+  {
+    throw except::Exception(
+        "xml::lite::Speed::buildTOC() called with only one of"
+        " parent path or current element."
+        " Only neither or both are allowed.");
+  }
+
+  if (element == NULL)
+  {
+    element = mRootNode;
+    // If we are starting at the top, clear out the old TOC before rebuilding.
+    mTOC.clear();
+  }
+
+  std::vector<xml::lite::Element *> childs = element->getChildren();
+  size_t sz = childs.size();
+  for (size_t ii = 0; ii < sz; ii++)
+  {
+    std::string lname = childs[ii]->getLocalName();
+    if (local_registry.count(lname) == 0)
+    {
+      local_registry[lname] = 1;
+    }
+    else
+    {
+      ++local_registry[lname];
+    }
+    std::string toc_key = parent+">"
+                          +lname+"&"
+                          +str::toString<size_t>(local_registry[lname]);
+    mTOC[toc_key] = childs[ii];
+    buildTOC(toc_key, childs[ii]);
+  }
+
+}
+//=================================================================
+std::string xml::lite::Speed::getData(const std::string &elname)
+{
+  return mTOC[elname]->getCharacterData();
+}
+//=================================================================
+void xml::lite::Speed::initRoot(const std::string &rootName)
+{
+  setRootElement(createElement(rootName,"",""));
+}
+//=================================================================
+xml::lite::Speed::KeyParts xml::lite::Speed::extractParts(const std::string &longname)
+{
+  xml::lite::Speed::KeyParts parts;
+  // make this name a proper name by ensuring zero-ordinal names have '&1' 
+  std::vector<std::string> chunks = str::split(longname, ">");
+
+  size_t ord_del_idx;
+  for(size_t ii = 0; ii<chunks.size(); ++ii)
+  { 
+    ord_del_idx = chunks[ii].find_last_of("&");
+    if (ord_del_idx == std::string::npos) chunks[ii] += "&1";
```

Comment:
Even one-liners always go on their own line with braces:

```
if (ord_del_idx == std::string::npos)
{
    chunks[ii] = "&1";
}
```

Also, I think you could skip having `ord_del_idx` at all and just put the `find_last_of` check on the left side of the condition... if you want to keep it, move the variable declaration inside the for loop since you don't need it elsewhere